### PR TITLE
Instruments, animation overhaul and whole mod cleanup

### PR DESCRIPTION
**Pull request by misternebula**

### CODE REVIEW

**Comment by Raicuparta** [comment]
🏃🏿‍♀️

**Comment by Raicuparta** [comment]
🎺

**Comment by MLGityaJtotheA** [comment]
🎻

**Comment by misternebula** [comment]
🎹

File: QSB/Animation/AnimationSync.cs [hunk 1]
```diff
@@ -38,30 +52,56 @@ private void OnDestroy()
             _playerController.OnJump -= OnJump;
             _playerController.OnBecomeGrounded -= OnBecomeGrounded;
             _playerController.OnBecomeUngrounded -= OnBecomeUngrounded;
-            GlobalMessenger.RemoveListener(EventNames.SuitUp, OnSuitUp);
-            GlobalMessenger.RemoveListener(EventNames.RemoveSuit, OnSuitDown);
+
+            QSBSceneManager.OnSceneLoaded -= (OWScene scene, bool universe) => LoadControllers(universe);
+        }
+
+        private void LoadControllers(bool universe)
```
Comment (amazingalek) [inline QSB/Animation/AnimationSync.cs:59]: just don't call this if not in universe?

File: QSB/Animation/AnimationSync.cs [hunk 2]
```diff
@@ -38,30 +52,56 @@ private void OnDestroy()
...
+            _gabbroController = gabbroRoot.GetComponent<Animator>().runtimeAnimatorController;
+            var feldsparRoot = GameObject.Find("Traveller_HEA_Feldspar_ANIM_Talking");
+            _feldsparController = feldsparRoot.GetComponent<Animator>().runtimeAnimatorController;
         }
 
         private void InitCommon(Transform body)
         {
+            if (QSBSceneManager.IsInUniverse)
+            {
+                LoadControllers(true);
```
Comment (amazingalek) [inline QSB/Animation/AnimationSync.cs:79]: ah, so it's true anyway :D

File: QSB/Animation/AnimationSync.cs [hunk 3]
```diff
@@ -112,55 +144,30 @@ private void InitCrouchSync()
...
-        private void OnBecomeUngrounded() => SendTrigger(AnimTrigger.Ungrounded);
-
-        private void OnSuitUp() => SendTrigger(AnimTrigger.SuitUp);
-        private void OnSuitDown() => SendTrigger(AnimTrigger.SuitDown);
+        private void OnJump() => _netAnim.SetTrigger("Jump");
+        private void OnBecomeGrounded() => _netAnim.SetTrigger("Grounded");
+        private void OnBecomeUngrounded() => _netAnim.SetTrigger("Ungrounded");
 
-        public void SendTrigger(AnimTrigger trigger, float value = 0)
+        public void SendCrouch(float value = 0)
```
Comment (amazingalek) [inline QSB/Animation/AnimationSync.cs:151]: Crouching should ofc be extracted to new class, but you don't *have* to do that in this PR ;)

File: QSB/Animation/Events/QSBAnimationMessage.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            writer.WritePackedUInt32((uint)stateHash);
+            writer.Write(normalizedTime);
+            if (parameters == null)
+            {
+                writer.WriteBytesAndSize(parameters, 0);
+            }
+            else
+            {
+                writer.WriteBytesAndSize(parameters, parameters.Length);
+            }
```
Comment (amazingalek) [inline QSB/Animation/Events/QSBAnimationMessage.cs:32]: wut :P

File: QSB/Animation/QSBNetworkAnimator.cs [hunk 5]
```diff
@@ -0,0 +1,375 @@
+﻿using QSB.Animation.Events;
+using UnityEngine;
+using UnityEngine.Networking;
+
+namespace QSB.Animation
+{
+    // Cleaned up unity code. UNET is so broken I gave up and fixed it myself.
```
Comment (amazingalek) [inline QSB/Animation/QSBNetworkAnimator.cs:7]: That's great! :)

For later, we could check out https://github.com/vis2k/Mirror, it's a community edition of UNet, cleaned up and allegedly with fewer bugs.

File: QSB/DeathSync/DeathPatches.cs [hunk 6]
```diff
@@ -25,11 +27,10 @@ public static void BroadcastDeath(DeathType deathType)
             GlobalMessenger<DeathType>.FireEvent(EventNames.QSBPlayerDeath, deathType);
         }
 
-        public static void DoPatches()
+        public override void DoPatches()
```
Comment (amazingalek) [inline QSB/DeathSync/DeathPatches.cs:30]: love it

File: QSB/ElevatorSync/QSBElevator.cs [hunk 7]
```diff
@@ -30,15 +30,15 @@ private void InitValues()
             _owAudioSourceLP = _elevator.GetValue<OWAudioSource>("_owAudioSourceLP");
         }
 
-        public void RemoteCall(ElevatorDirection direction)
+        public void RemoteCall(bool direction)
```
Comment (amazingalek) [inline QSB/ElevatorSync/QSBElevator.cs:33]: now it's not clear what true and false means. maybe rename to `isGoingUp`?

File: QSB/ElevatorSync/QSBElevator.cs [hunk 8]
```diff
@@ -30,15 +30,15 @@ private void InitValues()
...
         {
             SetDirection(direction);
             RemoteStartLift();
         }
 
-        private void SetDirection(ElevatorDirection direction)
+        private void SetDirection(bool direction)
         {
-            var isGoingUp = direction == ElevatorDirection.Up;
+            var isGoingUp = direction;
```
Comment (amazingalek) [inline QSB/ElevatorSync/QSBElevator.cs:41]: if you do the above suggestion, this line can ofc be removed :)

File: QSB/EventsCore/QSBEvent.cs [hunk 9]
```diff
@@ -1,8 +1,9 @@
 ﻿using QSB.Messaging;
+using QSB.Player;
 using QSB.TransformSync;
 using UnityEngine.Networking;
 
-namespace QSB.Events
+namespace QSB.EventsCore
```
Comment (amazingalek) [inline QSB/EventsCore/QSBEvent.cs:6]: not sure what the "Core" stuff is, but how about calling the namespace `QSB.Events.Core`? That's more "conventional"

File: QSB/Instruments/InstrumentsManager.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+        private void Awake()
+        {
+            Instance = this;
+            gameObject.AddComponent<CameraManager>();
+        }
+
+        private void Update()
+        {
+            if (Input.GetKeyDown(KeyCode.Keypad9))
```
Comment (amazingalek) [inline QSB/Instruments/InstrumentsManager.cs:22]: for debugging? 
the if could be inverted, and the key could be extracted to a constant.

File: QSB/Instruments/QSBCamera/CameraController.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
+﻿using UnityEngine;
+
+namespace QSB.Instruments.QSBCamera
+{
+    class CameraController : MonoBehaviour
```
Comment (amazingalek) [inline QSB/Instruments/QSBCamera/CameraController.cs:5]: is this taken from FreeCam, or is it something else? :D

File: QSB/MessagesCore/EnumMessage.cs [hunk 12]
```diff
@@ -1,22 +1,22 @@
 ﻿using QSB.Messaging;
 using UnityEngine.Networking;
 
-namespace QSB.DeathSync
+namespace QSB.MessagesCore
 {
-    public class PlayerDeathMessage : PlayerMessage
+    public class EnumMessage<T> : PlayerMessage
```
Comment (amazingalek) [inline QSB/MessagesCore/EnumMessage.cs:6]: yeeeeeees, generics <3

File: QSB/OrbSync/OrbPatches.cs [hunk 13]
```diff
@@ -1,11 +1,13 @@
-﻿using QSB.Events;
+﻿using QSB.EventsCore;
 using QSB.WorldSync;
 using UnityEngine;
 
 namespace QSB.OrbSync
 {
-    public static class OrbPatches
+    public class OrbPatches : QSBPatch
```
Comment (amazingalek) [inline QSB/OrbSync/OrbPatches.cs:7]: i thought patch classes also had to be static for Harmony to work. if this works, it's great!

File: QSB/QSB.csproj [hunk 14]
```diff
@@ -37,6 +37,10 @@
     <Reference Include="Assembly-CSharp">
       <HintPath>$(GameDir)\OuterWilds_Data\Managed\Assembly-CSharp.dll</HintPath>
     </Reference>
+    <Reference Include="Assembly-CSharp-firstpass, Version=0.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (amazingalek) [inline QSB/QSB.csproj:40]: what's this for?

File: QSB/QSB.csproj.user [hunk 15]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <GameDir>D:\EpicGames\OuterWilds</GameDir>
+    <GameDir>E:\Epic\Epic Games\OuterWilds</GameDir>
     <OwmlDir>C:\Users\Henry\AppData\Roaming\OuterWildsModManager\OWML</OwmlDir>
```
Comment (amazingalek) [inline QSB/QSB.csproj.user:5]: love that this is still in here :P

File: QSB/QSBNetworkManager.cs [hunk 16]
```diff
@@ -113,12 +104,14 @@ public override void OnStartServer()
...
-                    Destroy(item);
-                }
             }
+
+            NetworkServer.UnregisterHandler(40);
+            NetworkServer.UnregisterHandler(41);
+            NetworkServer.UnregisterHandler(42);
+            NetworkServer.RegisterHandler(40, new NetworkMessageDelegate(QSBNetworkAnimator.OnAnimationServerMessage));
+            NetworkServer.RegisterHandler(41, new NetworkMessageDelegate(QSBNetworkAnimator.OnAnimationParametersServerMessage));
+            NetworkServer.RegisterHandler(42, new NetworkMessageDelegate(QSBNetworkAnimator.OnAnimationTriggerServerMessage));
```
Comment (amazingalek) [inline QSB/QSBNetworkManager.cs:114]: 🤔
these magic numbers are in an enum somewhere, right?

Comment (amazingalek) [reply QSB/QSBNetworkManager.cs:114]: Yep - `eventType + MsgType.Highest`

File: QSB/QSBNetworkManager.cs [hunk 17]
```diff
@@ -136,50 +129,54 @@ public override void OnStartServer()
...
-            if (NetworkClient.active && !NetworkServer.active)
+            if (!NetworkServer.localClientActive)
             {
-                GeyserManager.Instance.EmptyUpdate();
-                WakeUpPatches.AddPatches();
+                QSBPatchManager.DoPatchType(QSBPatchTypes.OnNonServerClientConnect);
+                singleton.client.UnregisterHandler(40);
+                singleton.client.UnregisterHandler(41);
+                singleton.client.RegisterHandlerSafe(40, new NetworkMessageDelegate(QSBNetworkAnimator.OnAnimationClientMessage));
+                singleton.client.RegisterHandlerSafe(41, new NetworkMessageDelegate(QSBNetworkAnimator.OnAnimationParametersClientMessage));
```
Comment (amazingalek) [inline QSB/QSBNetworkManager.cs:145]: Should probably extract a ReRegisterEvent method, maybe somethinbg like this:
```
private void ReRegisterEvent(QsbEvent ev, Action action) 
{
  var id = ev + MsgType.Hightest + 1;
  singleton.client.UnregisterHandler(id);
  singleton.client.RegisterHandlerSafe(id, action);
}
```

File: QSB/Instruments/QSBCamera/CameraController.cs (hunk 11 above)
Comment (misternebula) [reply QSB/Instruments/QSBCamera/CameraController.cs:5]: the input stuff and rotation is taken from playercameracontroller and tweaked slightly, the position stuff is custom

File: QSB/QSB.csproj (hunk 14 above)
Comment (misternebula) [reply QSB/QSB.csproj:40]: post processing

File: QSB/QSBNetworkManager.cs (hunk 16 above)
Comment (misternebula) [reply QSB/QSBNetworkManager.cs:114]: these aren't the highest numbers though. these are in built unity messages that i'm taking away from NetworkAnimator and adding to QSBNetworkAnimator

File: QSB/OrbSync/OrbPatches.cs (hunk 13 above)
Comment (misternebula) [reply QSB/OrbSync/OrbPatches.cs:7]: i think only patch methods have to be static

**Pull request merged**